### DISCUSSION TITLE
Feature Request: Two factor authentication

### DISCUSSION THREAD

**Discussion by martijnhartlief** [postalserver/postal#1480] [Feature suggestions]
Is it possible to implement and optionally require 2FA for users?

**Comment by willpower232** [comment]
Its definitely possible but it will rely on someone who knows rails well enough to put in the work for it.

I'd definitely like to see it included but its a bit over my head at the minute.

**Comment by c-nv-s** [comment]
> Is it possible to implement and optionally require 2FA for users?

This would definitely be amazing

**Comment by RyankePars** [comment]
> Is it possible to implement and optionally require 2FA for users?

That would be great!

**Comment by brywhi** [comment]
Upvoting this, I might have time to contribute to this later

**Reply by willpower232** [reply]
Postal v2 has a user management area so this should be easier to implement now. Just need a good gem for it I guess :sweat_smile:

**Comment by WEBDIENSTE** [comment]
+1

**Comment by ivansalloum** [comment]
You can setup Caddy to use BasicAuth, which I've explained in my [guide](https://ivansalloum.com/setting-up-postal-as-an-smtp-server/#securing-the-web-interface). This adds an extra layer of security.

**Reply by rakurtz** [reply]
Hey Ivan, thanks again for your awesome article / guide. Just one thought about the caddy basic auth solution: Maybe the caddy config should allow the api-route without basic auth. Anyone using the api to send emails needs bypass the caddy basic-auth layer.

Ideally the api route is restricted to only accept certain IPs since in many situations it will only be a certain set of servers (with their respective IPs) using the Postal send-api.
